### PR TITLE
grpc: fixing timeout migration bug

### LINKED ISSUE #14909: max_stream_duration does not match deprecated max_grpc_timeout behaviour
*Title*: `max_stream_duration` does not match deprecated `max_grpc_timeout` behaviour

*Description*:

Hello, after the deprecation message we've tried to upgrade our config to move from `max_grpc_timeout` to the new options implemented in #12578, but it seems they behave differently. When  `max_grpc_timeout: 0s` is set everything works right and grpc streams do not timeout, but changing that to the new options stream requests do timeout:
```
 max_stream_duration:
    max_stream_duration: 0s
    grpc_timeout_header_max: 0s
```

Is there something that I'm missing?

Envoy version: `1.16.2`

### PR DESCRIPTION
**Pull request by alyssawilk**
Fixing an issue with gRPC timeouts where using the new timeouts didn't disable old timeout (including defaults).

Risk Level: medium (changes behavior for folks using new config)
Testing: new unit tests, integration tests
Docs Changes: n/a
Release Notes: inline
Fixes #14909

### CODE REVIEW

File: source/common/http/conn_manager_impl.cc [hunk 1]
```diff
@@ -1218,10 +1218,20 @@ void ConnectionManagerImpl::ActiveStream::refreshDurationTimeout() {
...
+        request_headers_->removeGrpcTimeout();
+      }
     }
     return;
   }
 
+  // Set the header timeout before doing used-time adjustements.
+  // This may result in the upstream not getting the latest results, but also
+  // avoids every request getting a custom timeout based on envoy think time.
+  if (route->usingNewTimeouts() && Grpc::Common::isGrpcRequestHeaders(*request_headers_)) {
```
Comment (alyssawilk) [inline source/common/http/conn_manager_impl.cc:1231]: Also noteworthy, the new code does *not* set the grpc-timeout like the old code does.

The old code would set the grpc-timeout to whatever was set in Envoy, but I'm not confident that's the correct behavior.
Imagine here selecting one route which took the header based timeout of 4s and lowered to 2s, then a route change to a route with no timeout.  I think functionally we'd want to set the grpc-timeout header iff Envoy lowered the timeout, which would require latching the original downstream timeout.  Which I can go do if you think it's worth it, but given Envoy doesn't advertise any other timeouts I'm wondering if we can just leave as-is?

**Comment by alyssawilk** [comment]
Mac timeout is known, so PTAL when you get a chance.

**Review by lizan — APPROVED** [review]
logic lgtm, can you check CI?

**Comment by alyssawilk** [comment]
yeah, looks like that quic breakage, I'll merge main.

**Pull request merged**